sql_export/sql_export.py

```diff
@@ -0,0 +1,142 @@
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import StringIO
+import base64
+import datetime
+import re
+from openerp import models, fields, api, _, exceptions
+from openerp.tools import DEFAULT_SERVER_DATETIME_FORMAT
+
+
+class SqlExport(models.Model):
+    _name = "sql.export"
+    _description = "SQL export"
+
+    PROHIBITED_WORDS = [
+        'delete',
+        'drop',
+        'insert',
+        'alter',
+        'truncate',
+        'execute',
+        'create',
+        'update'
+    ]
+
+    @api.multi
+    def _check_query_allowed(self):
+        for obj in self:
+            query = obj.query.lower()
+            for word in self.PROHIBITED_WORDS:
+                expr = r'\b%s\b' % word
+                is_not_safe = re.search(expr, query)
+                if is_not_safe:
+                    return False
+        return True
+
+    @api.model
+    def _get_editor_group(self):
+        ir_model_obj = self.env['ir.model.data']
+        return [ir_model_obj.xmlid_to_res_id(
+            'sql_export.group_sql_request_editor')]
+
+    name = fields.Char('Name', required=True)
+    query = fields.Text(
+        'Query',
+        required=True,
+        help="You can't use the following word : delete, drop, create, "
+             "insert, alter, truncate, execute, update")
+    copy_options = fields.Char(
+        'Copy Options',
+        required=True,
+        default="CSV HEADER DELIMITER ';'")
+    group_ids = fields.Many2many(
+        'res.groups',
+        'groups_sqlquery_rel',
+        'sql_id',
+        'group_id',
+        'Allowed Groups',
+        default=_get_editor_group)
+    user_ids = fields.Many2many(
+        'res.users',
+        'users_sqlquery_rel',
+        'sql_id',
+        'user_id',
+        'Allowed Users')
+
+    _constraints = [(_check_query_allowed,
+                     'The query you want make is not allowed : prohibited '
+                     'actions (%s)' % ', '.join(PROHIBITED_WORDS),
+                     ['query'])]
+
+    @api.multi
+    def export_sql_query(self):
+        for obj in self:
+            today = datetime.datetime.now()
+            today_tz = fields.Datetime.context_timestamp(
+                obj, today)
+            date = today_tz.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
+            output = StringIO.StringIO()
+            query = "COPY (" + obj.query + ")  TO STDOUT WITH " + \
+                    obj.copy_options
+            self.env.cr.copy_expert(query, output)
```

Comment:
I think there is no real risk because the query is executed with copy_expert and it does not accept 'update', 'drop', etc either.

But I agree with you, it can't hurt to be paranoid, I have added the rolllback statement.

Thank you for the review!